docs/configs/evok_configuration.md
# Evok configuration

Evok configuration is located in `/etc/evok/config.yaml`. The default configuration is installed by the debian package. To apply the configuration, it is necessary to restart Evok with the command `systemctl restart evok`.

## API settings

In this section you can configure address and port for APIs listening. This setting will be applied to protocols:

- [REST](../apis/rest.md)
- [JSON](../apis/json.md)
- [BULK](../apis/bulk.md)
- [RPC](../apis/rpc.md)
- [Webhook](../apis/webhook.md)
- [WebSocket](../apis/websocket.md)

!!! info

    More detailed settings of individual protocols can be found in their separate section.

### Websocket

- `enabled` - enables websocket API (`true` / `false`)
- `all_filtered` - all WebSocket requests will be subject to the filtering set by 'filter' (`true` / `false`)

### Webhook

- `enabled` - enables webhook notifications (`true` / `false`)
- `address` - address (with port) to which notifications should be sent
- `device_mask` - list of devices to notify on (written as a JSON list, same format as `address`)
- `complex_events` - Evok will send POST requests with the same data as WebSocket, rather than an empty GET request

## Hardware configuration

The hardware configuration is represented by a device tree with this structure:

```yaml
comm_channels:
    <bus_name>:
        type: <bus_type>
        <bus specific settings>: <specific parameters>
        devices:
            <device_name>:
                slave-id: <slave_id>
                model: <model_id>
                scan_frequency: <scan_frequency>
```

### Bus configuration

- *<bus_name\>* - Your choice, but has to be unique
- `type` options:
    - `MODBUSTCP`
        - `hostname` - hostname of the Modbus server
        - `port` - port of the Modbus server
    - `MODBUSRTU`
        - `port` - path to the Modbus device
        - `boudrate` - baudrate of the Modbus device
        - `parity` - parity of the Modbus device (`N` / `E` / `O`)
    - `OWBUS`
        - `interval` - interval of values updating
        - `scan_interval` - new devices will be automatically assigned
        - `owpower` - [Circuit](../circuit.md) of owpower device (for restarting bus; optional parameter)

### Device configuration

- *<device_name\>*: the device will be available in the API under this name. Has to be unique.

#### MODBUSTCP & MODBUSRTU

- `model_id` - assigns a Modbus register map (examples: `xS51`, `xS11`), see [hw_definitions](./hw_definitions.md).
- `slave_id` - slave address or unit-ID of the Modbus device.
- `scan_frequency` - an optional parameter, determines how often values are read from the device (Default value is 50).

#### OWBUS

- `type` - 1-Wire sensor type, options: [`DS18B20`, `DS18S20`, `DS2438`, `DS2408`, `DS2406`, `DS2404`, `DS2413`]
- `address` - 1-Wire device address

!!! Note
    It is better to use automatic device search, rather than defining devices manually.

### Examples

Every definition must be in `comm_channels` section.

#### Modbus RTU device

```yaml title="xS51 on /dev/ttyNS0"
RS485_1:
  type: MODBUSRTU
  port: /dev/ttyNS0
  baudrate: 19200
  parity: 'N'
  devices:
    xS51:
      slave-id: 1
      model: xS51
```

#### Modbus TCP device

```yaml title="IAQ on TCP"
TCP_EXT:
  type: MODBUSTCP
  hostname: 192.168.0.54
  port: 502
  devices:
    myIAQ:
      slave-id: 1
      model: IAQ
```

#### 1-Wire device

```yaml title="1-Wire thermometer"
TEMPM:
  type: OWBUS
  interval: 10
  scan_interval: 60
  owpower: 1
```

## Autogen

If the Debian package `unipi-os-configurator` is installed,
Evok can automatically create the hardware configuration for the running device,
but it works only for Unipi controllers.
You can enable this feature with `autogen: true` in config.
If this feature is enabled, Evok includes the file `/etc/evok/autogen.yaml`.
This file contains the hardware configuration of the running device.
`unipi-os-configurator` generates this file if a hardware change has been detected.
You can force the creation of this file using this command:

```bash
/opt/unipi/tools/os-configurator -f
```

### Autogen rules
 - The Bus is always named `LOCAL_TCP`.
 - A device_info section is generated to describe the device. This information is based on `unipiid`.
 - The device name is generated based on the `slave-id`. In standard Unipi controllers it is the same as the section number.
 - The OWFS section is generated only if the device supports 1-Wire and the `owserver` package is installed.
   - The `owpower` parameter is defined only if Unipi controller supports this feature.

```yaml title="Autogen example"
comm_channels:
  LOCAL_TCP:
    type: MODBUSTCP
    hostname: 127.0.0.1
    port: 502
    device_info:
      family: Neuron
      model: L533
      sn: 0
      board_count: 3
    devices:
      1:
        slave-id: 1
        model: 00
      2:
        slave-id: 2
        model: 13
      3:
        slave-id: 3
        model: 13
  OWFS:
    type: OWBUS
    interval: 10
    scan_interval: 60
    owpower: 1
```


debian/changelog
evok (2.0.7i) unstable; urgency=medium

  * Fixed WebSocket notification for relays and RS485 extension LEDs
  * Fixed PWM frequency computation not being correctly reapplied when restarting EVOK
  * Allowed setting PWM duty cycle to 0; this will turn off PWM mode after restart due to the hardware not being able to distinguish between a duty cycle of 0 and PWM turned off. See our ModBus map for details of the underlying implementation.

 -- Tomas Knot <knot@unipi.technology>  Wed, 20 Feb 2018 19:41:12 +0200

evok (2.0.7h) unstable; urgency=medium

  * Fixed WebSocket notification for secondary analog inputs

 -- Tomas Knot <knot@unipi.technology>  Tue, 05 Feb 2018 15:31:12 +0200

evok (2.0.7g) unstable; urgency=medium

  * Removed unnecessary errors in the log

 -- Tomas Knot <knot@unipi.technology>  Mon, 22 Jan 2018 15:51:08 +0200

evok (2.0.7f) unstable; urgency=medium

  * Fixed issues with custom device parsing

 -- Tomas Knot <knot@unipi.technology>  Mon, 22 Jan 2018 14:44:08 +0200

evok (2.0.7e) unstable; urgency=medium

  * Added the ability to listen for analog input events to the websocket filter
  * Fixed setting watchdog via the JSON endpoint
  * Simplified default NGINX configuration

 -- Tomas Knot <knot@unipi.technology>  Mon, 22 Jan 2018 10:54:08 +0200

evok (2.0.7d) unstable; urgency=medium

  * Added an option to turn off filtering for function -all- websocket calls

 -- Tomas Knot <knot@unipi.technology>  Mon, 10 Dec 2018 10:54:08 +0200

evok (2.0.7b) unstable; urgency=medium

  * Added separation between input and holding registers.

 -- Tomas Knot <knot@unipi.technology>  Mon, 10 Sep 2018 11:37:08 +0200

evok (2.0.7a) unstable; urgency=medium

  * Initial Release.

 -- Tomas Knot <knot@unipi.technology>  Thu, 29 Mar 2018 15:51:08 +0200


evok/config.py
import os
from typing import List, Dict, Union
from tornado.ioloop import IOLoop

from .modbus_unipi import EvokModbusSerialClient, EvokModbusTcpClient
from .modbus_slave import ModbusSlave
from . import owdevice

import yaml
from .devices import *


class EvokConfigError(Exception):
    pass


class HWDict:
    def __init__(self, dir_paths: List[str] = None, paths: List[str] = None):
        """
        :param dir_paths: path to dir for load
        :param paths: paths to config files
        """
        self.definitions: Dict[str, List] = {}
        scope = list()
        if dir_paths is not None:
            for dp in dir_paths:
                scope.extend([dp + f for f in os.listdir(dp)])
        if paths is not None:
            scope.extend(paths)
        if scope is None or len(scope) == 0:
            raise ValueError(f"HWDict: no scope!")
        for file_path in scope:
            if file_path.endswith(".yaml") and os.path.isfile(file_path):
                file_name = file_path.split("/")[-1].replace(".yaml", "")
                with open(file_path, 'r') as yfile:
                    ydata = yaml.load(yfile, Loader=yaml.SafeLoader)
                    if ydata is None:
                        logger.warning(f"Empty Definition file '{file_path}'! skipping...")
                        continue
                    self.definitions[file_name] = ydata
                    logger.debug(f"YAML Definition loaded: {file_path}, definition count {len(self.definitions) - 1}")


class OWSensorDevice:
    def __init__(self, sensor_dev):
        self.sensor_dev = sensor_dev
        self.circuit = sensor_dev.circuit

    def full(self):
        return self.sensor_dev.full()


class TcpBusDevice:
    def __init__(self, circuit: str, bus_driver: EvokModbusTcpClient):
        self.bus_driver = bus_driver
        self.circuit = circuit

    def switch_to_async(self, loop: IOLoop):
        loop.add_callback(lambda: self.bus_driver.connect())


class SerialBusDevice:
    def __init__(self, circuit: str,  bus_driver: EvokModbusSerialClient):
        self.bus_driver = bus_driver
        self.circuit = circuit

    def switch_to_async(self, loop: IOLoop):
        self.bus_driver.ioloop = loop
        loop.add_callback(lambda: self.bus_driver.connect())


class DeviceInfo:
    def __init__(self, name: str, family: str, model: str, sn: Union[None, int], board_count: int):
        """
        :param family: [Neuron, Patron, UNIPI1, Iris]
        :param model: [S103, M533, ...]
        :param sn: serial number
        :param board_count: number of boards
        """
        self.family: str = family
        self.model: str = model
        self.sn: Union[None, int] = sn
        self.board_count: int = board_count
        self.circuit: str = f"{name}"

    def full(self):
        return {
            'dev': "device_info",
            'family': self.family,
            'model': self.model,
            'sn': self.sn,
            'board_count': self.board_count,
            'circuit': self.circuit,
        }


class EvokConfig:

    def __init__(self, conf_dir_path: str):
        data = self.__get_final_conf(scope=[conf_dir_path+'/config.yaml'])
        self.comm_channels: dict = self.__get_comm_channels(data)
        self.apis: dict = self.__get_apis_conf(data)
        self.logging: dict = self.__get_logging_conf(data)

    def __merge_data(self, source: dict, append: dict):
        for key in append:
            if key in source and type(source[key]) == dict and type(append[key]) == dict:
                source[key] = self.__merge_data(source[key], append[key])
            else:
                source[key] = append[key]
        return source

    def __get_final_conf(self, conf_dir_path: Union[None, str] = None,
                         scope: Union[None, List[str]] = None,
                         check_autogen: bool = True) -> dict:
        if scope is None:
            files = os.listdir(conf_dir_path)
            if 'config.yaml' not in files:
                raise EvokConfigError(f"Missing 'config.yaml' in evok configuration directory ({conf_dir_path})")
            scope = files
        final_conf = {}
        for path in scope:
            try:
                with open(path, 'r') as f:
                    ydata: dict = yaml.load(f, Loader=yaml.Loader)
                self.__merge_data(final_conf, ydata)
            except FileNotFoundError:
                logger.warning(f"Config file {path} not found!")
        if check_autogen and final_conf.get('autogen', False):
            return self.__get_final_conf(scope=['/etc/evok/autogen.yaml', *scope], check_autogen=False)
        return final_conf

    @staticmethod
    def __get_comm_channels(data: dict) -> dict:
        ret = {}
        if 'comm_channels' not in data:
            logger.warning("Section 'comm_channels' not in configuration!")
            return ret
        for name, value in data['comm_channels'].items():
            ret[name] = value
        return ret

    @staticmethod
    def __get_apis_conf(data: dict) -> dict:
        ret = {}
        if 'apis' not in data:
            logger.warning("Section 'apis' not in configuration!")
            return ret
        for name, value in data['apis'].items():
            ret[name] = value
        return ret

    @staticmethod
    def __get_logging_conf(data: dict) -> dict:
        ret = {}
        if 'logging' not in data:
            logger.warning("Section 'logging' not in configuration!")
            return ret
        for name, value in data['logging'].items():
            ret[name] = value
        return ret

    def configtojson(self):
        return self.main  # TODO: zkontrolovat!!

    def get_comm_channels(self) -> dict:
        return self.comm_channels

    def get_api(self, name: str) -> dict:
        if name not in self.apis:
            logging.warning(f"Api '{name}' not found")
            return {}
        return self.apis[name]


def hexint(value):
    if value.startswith('0x'):
        return int(value[2:], 16)
    return int(value)


def create_devices(evok_config: EvokConfig, hw_dict):
    for bus_name, bus_data in evok_config.get_comm_channels().items():
        bus_data: dict
        if not bus_data.get("enabled", True):
            logger.info(f"Skipping disabled bus '{bus_name}'")
            continue
        bus_type = bus_data['type']

        bus = None
        bus_device_info: Union[None, DeviceInfo] = None
        if bus_type == 'OWFS':
            interval = bus_data.get("interval", 60)
            scan_interval = bus_data.get("scan_interval", 300)
            owpower = bus_data.get("owpower", None)

            circuit = bus_name
            bus = owdevice.OwBusDriver(circuit, interval=interval, scan_interval=scan_interval,
                                       owpower_circuit=owpower)
            Devices.register_device(OWBUS, bus)

        elif bus_type == 'MODBUSTCP':
            modbus_server = bus_data.get("hostname", "127.0.0.1")
            modbus_port = bus_data.get("port", 502)
            bus_driver = EvokModbusTcpClient(host=modbus_server, port=modbus_port)
            bus = TcpBusDevice(circuit=bus_name, bus_driver=bus_driver)
            Devices.register_device(TCPBUS, bus)

        elif bus_type == "MODBUSRTU":
            serial_port = bus_data["port"]
            serial_baud_rate = bus_data.get("baudrate", 19200)
            serial_parity = bus_data.get("parity", 'N')
            serial_stopbits = bus_data.get("stopbits", 1)
            bus_driver = EvokModbusSerialClient(port=serial_port, baudrate=serial_baud_rate, parity=serial_parity,
                                                stopbits=serial_stopbits, timeout=0.5)
            bus = SerialBusDevice(circuit=bus_name, bus_driver=bus_driver)
            Devices.register_device(SERIALBUS, bus)

        if bus is not None:
            bus_device_info_data = bus_data.get("device_info", None)  # noqa
            if bus_device_info_data is not None:
                bus_device_info_data: dict
                family = bus_device_info_data.get("family", 'unknown')
                model = bus_device_info_data.get("model", 'unknown')
                sn = bus_device_info_data.get("sn", None)
                board_count = bus_device_info_data.get("board_count", 1)
                bus_device_info = DeviceInfo(name=model, family=family, model=model, sn=sn, board_count=board_count)
                Devices.register_device(DEVICE_INFO, bus_device_info)

        if 'devices' not in bus_data:
            logger.info(f"Creating bus '{bus_name}' with type '{bus_type}'.")
            continue

        logger.info(f"Creating bus '{bus_name}' with type '{bus_type}' with devices.")
        for device_name, device_data in bus_data['devices'].items():
            if not device_data.get("enabled", True):
                logger.info(f"^ Skipping disabled device '{device_name}'")
                continue
            logger.info(f"^ Creating device '{device_name}' with type '{bus_type}'")
            try:
                if bus_type == 'OWBUS':
                    ow_type = device_data.get("type")
                    address = device_data.get("address")
                    interval = device_data.getintdef("interval", 15)

                    circuit = device_name
                    sensor = owdevice.MySensorFabric(address, ow_type, bus, interval=interval, circuit=circuit,
                                                     is_static=True)
                    if sensor is not None:
                        sensor = OWSensorDevice(sensor)
                    Devices.register_device(SENSOR, sensor)

                elif bus_type in ['MODBUSTCP', 'MODBUSRTU']:
                    slave_id = device_data.get("slave-id", 1)
                    scanfreq = device_data.get("scan_frequency", 50)
                    scan_enabled = device_data.get("scan_enabled", True)
                    device_model = device_data["model"]
                    circuit = f"{device_name}"
                    major_group = device_name

                    slave = ModbusSlave(bus.bus_driver, circuit, evok_config, scanfreq, scan_enabled,
                                        hw_dict, device_model=device_model, slave_id=slave_id,
                                        major_group=major_group)
                    Devices.register_device(MODBUS_SLAVE, slave)

                    if bus_device_info is None or "device_info" in device_data:
                        device_info = {'model': device_data.get("model", device_name)}
                        device_info.update(device_data.get("device_info", {}))
                        family = device_info.get("family", 'unknown')
                        model = device_info.get("model", 'unknown')
                        sn = device_info.get("sn", None)
                        board_count = device_info.get("board_count", 1)
                        if model[:2].lower() in ['xs', 'xm', 'xl', 'xg'] and family == 'unknown':
                            family = 'Extension'
                        Devices.register_device(DEVICE_INFO,
                                                DeviceInfo(name=device_name, family=family, model=model, sn=sn,
                                                           board_count=board_count))

                else:
                    logger.error(f"Unknown bus type: '{bus_type}'! skipping...")
                    if logger.level == logging.DEBUG:
                        traceback.print_exc()

            except Exception as E:
                logger.exception(f"Error in config section '{bus_type}:{device_name}' - {str(E)}")


def load_aliases(path):
    alias_dicts = list(HWDict(paths=[path]).definitions.values())
    # HWDict returns List(), take only first item
    alias_conf = alias_dicts[0] if len(alias_dicts) > 0 else dict()
    version = str(alias_conf.get("version", "1.0"))

    version = alias_conf.get("version", None)
    if version == "1.0":
        # transform array to dict and rename dev_type -> devtype if version 1.0
        result = dict(((rec["name"], {"circuit": rec.get("circuit", None), "devtype": rec.get("dev_type", None)})\
                       for rec in alias_conf.get("aliases", {})\
                       if rec.get("name", None) is not None))
    elif version == "2.0":
        result = alias_conf.get("aliases", {})
    else:
        result = {}
    Devices.aliases = Aliases(result)
    logger.debug(f"Load aliases with {result}")


# don't call it directly in asyn loop -- block
def save_aliases(alias_dict, path):
    try:
        logger.info(f"Saving alias file {path}")
        with open(path, 'w+') as yfile:
            yfile.write(yaml.dump({"version": "2.0", "aliases": alias_dict}))
        os.system('sync')
    except Exception as E:
        logger.exception(str(E))
